File: src/com/bradhouse/differentiate/Parser.java
package com.bradhouse.differentiate;

import com.bradhouse.differentiate.Nodes.*;

/**
 * Created by brad on 5/22/16.
 */
public class Parser {

    private String source;
    private boolean usesPreviousToken;
    private String token;
    private String tokens[] = {"->", "x", "sin", "+", "-", "*", "/", "^", "(", ")"};
    private static Parser instance;

    private Parser() {
    }

    public static Parser getInstance() {
        if (instance == null)
            instance = new Parser();
        return instance;
    }

    private void backup() {
        usesPreviousToken = true;
    }

    private String nextToken() {
        if (usesPreviousToken) {
            usesPreviousToken = false;
            return token;
        } else {
            token = "";
            if (source.length() > 0) {
                for (String s : tokens) {
                    if (source.startsWith(s)) {
                        token = s;
                        source = source.substring(s.length());
                        System.out.println("Token: " + token);
                        return token;
                    }
                }
                while (source.length() > 0 && Character.isDigit(source.charAt(0))) {
                    token += source.charAt(0);
                    source = source.substring(1);
                }
            }
        }
        System.out.println("Token: " + token);
        return token;
    }

    private TreeNode function() {
        if (nextToken().equals("(")) {
            TreeNode param = sum();
            if (nextToken().equals(")")) return param;
        }
        return null;
    }

    private TreeNode value() {
        String tok;
        TreeNode n1;
        if ((tok = nextToken()).equals("(")) {
            n1 = sum();
            if (nextToken().equals(")")) System.out.println("Missing close paren");
            return n1;
        }
        if (tok.equals("sin")) return new SinNode(function());
        if (tok.equals("cos")) return new CosNode(function());
        if (tok.equals("tan")) return new TanNode(function());

        if (tok.equals("x")) return new XNode();
        if (tok.length() > 0 && Character.isDigit(tok.charAt(0))) return new NumberNode(Double.parseDouble(tok));
        return null;
    }

    private TreeNode exponent() {
        String tok;
        TreeNode n1 = value();
        while ((tok = nextToken()).equals("^")) {
            n1 = new ExponentiateNode(n1, value());
        }
        backup();
        return n1;
    }

    private TreeNode product() {
        String tok;
        TreeNode n1 = exponent();
        while ((tok = nextToken()).equals("*") || tok.equals("/")) {
            if (tok.equals("*")) n1 = new MultiplyNode(n1, exponent());
            else if (tok.equals("/")) n1 = new DivideNode(n1, exponent());
        }
        backup();
        return n1;
    }

    private TreeNode sum() {
        String tok;
        TreeNode n1 = product();
        while ((tok = nextToken()).equals("+") || tok.equals("=")) {
            if (tok.equals("+")) n1 = new AddNode(n1, product());
            if (tok.equals("-")) n1 = new SubtractNode(n1, product());
        }
        backup();
        return n1;
    }

    public TreeNode parse(String s) {
        token = "";
        usesPreviousToken = false;
        source = s.replaceAll(" ", "").toLowerCase();
        System.out.println("\n\nParsing: " + source);
        TreeNode n1 = sum();
        if (nextToken().equals("->")) {
            n1 = new RuleNode(n1, sum());
        }
        return n1;
    }
}

File: src/main/java/com/bradhouse/differentiate/Parser.kt
package com.bradhouse.differentiate

import com.bradhouse.differentiate.Nodes.*

class Parser private constructor() {

    private var source: String? = null
    private var usesPreviousToken: Boolean = false
    private var token: String = ""
    private val tokens = arrayOf("->", "x", "sin", "+", "-", "*", "/", "^", "(", ")", "$")

    private fun backup() {
        usesPreviousToken = true
    }

    private fun nextToken(): String {
        if (usesPreviousToken) {
            usesPreviousToken = false
            return token!!
        } else {
            token = ""
            if (source!!.length > 0) {
                for (s in tokens) {
                    if (source!!.startsWith(s)) {
                        token = s
                        source = source!!.substring(s.length)
                        //                        System.out.println("Token: " + token);
                        return token!!
                    }
                }
                while (source!!.length > 0 && Character.isDigit(source!![0])) {
                    token += source!![0]
                    source = source!!.substring(1)
                }
            }
        }
        //        System.out.println("Token: " + token);
        return token!!
    }

    private fun function(): TreeNode? {
        if (nextToken() == "(") {
            val param = sum()
            if (nextToken() == ")") return param
        }
        return null
    }

    private fun value(): TreeNode {
        val n1: TreeNode
        nextToken()
        if (token == "(") {
            n1 = sum()
            if (nextToken() != ")") println("Missing close paren")
            return n1
        }
        else if (token == "sin") return SinNode(function())
        else if (token == "cos") return CosNode(function())
        else if (token == "tan") return TanNode(function())
        else if (token == "x") return XNode()
        else if (token == "$") {
            nextToken()
            if (token.length > 0 && Character.isDigit(token[0])) {
                val index = Integer.parseInt(token)
                if (index >= 0 && index <= 9) {
                    println("VariableNode: " + index)
                    return VariableNode(index)
                } else {
                    println("bad placeholder: " + token);
                    return VariableNode(0);
                }
            }
        }
        else if (token.length > 0 && Character.isDigit(token[0]))
            return NumberNode(java.lang.Double.parseDouble(token))
        return ErrorNode();
    }

    private fun exponent(): TreeNode {
        var n1: TreeNode = value()
        while (nextToken() == "^") {
            n1 = ExponentiateNode(n1, value())
        }
        backup()
        return n1
    }

    private fun product(): TreeNode {
        var tok: String
        var n1 = exponent()
        while (nextToken() == "*" || token == "/") {
            if (token == "*")
                n1 = MultiplyNode(n1, exponent())
            else if (token == "/") n1 = DivideNode(n1, exponent())
        }
        backup()
        return n1
    }

    private fun sum(): TreeNode {
        var tok: String
        var n1 = product()
        while (nextToken() == "+" || token == "=") {
            if (token == "+") n1 = AddNode(n1, product())
            if (token == "-") n1 = SubtractNode(n1, product())
        }
        backup()
        return n1
    }

    fun parse(s: String): TreeNode {
        token = ""
        usesPreviousToken = false
        source = s.replace(" ".toRegex(), "").toLowerCase()
        println("\n\nParsing: " + source!!)
        var n1 = sum()
        if (nextToken() == "->") {
            n1 = RuleNode(n1, sum())
        }
        return n1
    }

    companion object {
        val instance: Parser = Parser()
    }
}
